ballerina/graphql_client.bal

```diff
@@ -0,0 +1,183 @@
+                    return error ClientError("GraphQL Client Error", body = data);
+                }
+                return error ClientError("GraphQL Client Error", httpResponse);
+            } else {
+                map<json> responseMap = <map<json>> httpResponse;
+
+                if responseMap.hasKey("errors") {
+                    GraphQLError[] errors = check responseMap.get("errors").cloneWithType(GraphQLErrorArray);
+                    
+                    if (responseMap.hasKey("data") && !responseMap.hasKey("extensions")) {
+                        return error ServerError("GraphQL Server Error", data = responseMap.get("data"), errors = errors);
+                    } else if (responseMap.hasKey("extensions") && !responseMap.hasKey("data")) {
+                        map<json>? extensionsMap = 
+                            (responseMap.get("extensions") is ()) ? () : <map<json>> responseMap.get("extensions");
+                        return error ServerError("GraphQL Server Error", errors = errors, extensions = extensionsMap);
+                    } else if (responseMap.hasKey("data") && responseMap.hasKey("extensions")) {
+                        map<json>? extensionsMap = 
+                            (responseMap.get("extensions") is ()) ? () : <map<json>> responseMap.get("extensions") ;
+                        return error ServerError("GraphQL Server Error", data = responseMap.get("data"), errors = errors, 
+                            extensions = extensionsMap);
+                    } else {
+                        return error ServerError("GraphQL Server Error", errors = errors);
+                    }
+                } else {
+                    if targetType is typedesc<GenericResponse> {
+                        GenericResponse response = check responseMap.cloneWithType(targetType);
+                        return response;
+                    } else if targetType is typedesc<DataResponse> {
+                       json responseData = responseMap.get("data");
+                       if (responseMap.hasKey("extensions")) {
+                          responseData = check responseData.mergeJson({ "__extensions" : responseMap.get("extensions") });
+                       }
+                       DataResponse response = check responseData.cloneWithType(targetType);
+                       return response;
+                    } else if targetType is typedesc<record{}> {
+                        record{} response = check httpResponse.cloneWithType(targetType);
+                        return response;
+                    } else if targetType is typedesc<json> {
+                        json response = check httpResponse.cloneWithType(targetType);
+                        return response;
+                    } else {
+                        return error ClientError("GraphQL Client Error, Invalid binding type.");
+                    }
+                }
+            }
+        } on fail var e {
```

Comment:
What's the purpose of using a `do`-`on fail` block here? The code will return `ClientError` or a `ServerError` based on the scenario. There's no need to encapsulate it with another error, right?

Reply:
We use the `do-on fail` block to catch all the generic response processing errors and to return them as a `ClientError` which we renamed as a `RequestError`.